File: android/app/src/main/java/com/groundworkpro/app/AttendanceQueueStorePlugin.kt
package com.groundworkpro.app

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.io.File

// Durable storage for the offline attendance queue (Android).
//
// STATUS: REFERENCE IMPLEMENTATION — NOT YET WIRED INTO THE BUILD OR VERIFIED ON
// A PHYSICAL DEVICE. See docs/attendance-offline-sync.md. Until the plugin is
// registered and exercised on a real phone, the JS layer falls back to
// localStorage and offline durability across a device restart is NOT proven.
//
// Why not localStorage: the app is a remote-URL Capacitor shell, so the queue
// would live in WebView web storage — wiped by "clear cache" and evictable
// under storage pressure.
//
// The queue holds no secrets (the attendance credential lives in the Keystore
// via SecureAttendanceStore), so this is app-private INTERNAL storage for
// durability, not encrypted storage for secrecy. Internal storage is not
// world-readable, is excluded from ordinary backups by the shared
// GeofenceBroadcastReceiver's configuration, and is removed with the app.

@CapacitorPlugin(name = "AttendanceQueueStore")
class AttendanceQueueStorePlugin : Plugin() {
    companion object {
        const val FILE_NAME = "attendance-queue.json"

        // Shared with GeofenceBroadcastReceiver, which appends to the same queue
        // when a background transition cannot be POSTed immediately.
        fun queueFile(context: android.content.Context): File =
            File(context.filesDir, FILE_NAME)
    }

    @PluginMethod
    fun load(call: PluginCall) {
        val result = JSObject()
        val file = queueFile(context)
        result.put(
            "value",
            try {
                if (file.exists()) file.readText() else null
            } catch (e: Exception) {
                // A read failure must not break attendance: the JS layer falls
                // back to its localStorage mirror.
                null
            }
        )
        call.resolve(result)
    }

    @PluginMethod
    fun save(call: PluginCall) {
        val value = call.getString("value")
        if (value == null) {
            call.reject("value required")
            return
        }
        try {
            // Write-then-rename so a kill mid-write cannot leave a truncated
            // queue behind.
            val target = queueFile(context)
            val temp = File(context.filesDir, "$FILE_NAME.tmp")
            temp.writeText(value)
            if (!temp.renameTo(target)) {
                target.writeText(value)
                temp.delete()
            }
            call.resolve()
        } catch (e: Exception) {
            call.reject("Failed to persist attendance queue: ${e.message}")
        }
    }

    @PluginMethod
    fun clear(call: PluginCall) {
        try {
            val file = queueFile(context)
            if (file.exists()) file.delete()
            call.resolve()
        } catch (e: Exception) {
            call.reject("Failed to clear attendance queue: ${e.message}")
        }
    }
}

File: ios/App/App/AttendanceQueueStorePlugin.swift
import Foundation
import Capacitor

// Durable storage for the offline attendance queue (iOS).
//
// STATUS: REFERENCE IMPLEMENTATION — NOT YET WIRED INTO THE BUILD OR VERIFIED ON
// A PHYSICAL DEVICE. See docs/attendance-offline-sync.md. Until the plugin is
// registered in the Xcode project and exercised on a real phone, the JS layer
// falls back to localStorage and offline durability across a device restart is
// NOT proven.
//
// Why not localStorage: the app is a remote-URL Capacitor shell, so the queue
// would live in WKWebView web storage — which iOS may evict under storage
// pressure and which is cleared by ordinary web-data resets. Attendance that
// survives being offline but not being backgrounded for two days is not durable.
//
// Why not the Keychain: the queue holds no secrets (the attendance credential
// lives in the Keychain via SecureAttendanceStore). This is about DURABILITY.
// The file lives in Application Support — inside the app container, not
// user-visible, and removed with the app — with
// `.completeUntilFirstUserAuthentication` data protection so the native
// geofence handler can still read and append to it after a device restart but
// before the user has unlocked the phone.

@objc(AttendanceQueueStorePlugin)
public class AttendanceQueueStorePlugin: CAPPlugin {
    // Shared with the native geofence handler, which appends to the same queue
    // when a background transition cannot be POSTed immediately.
    static let fileName = "attendance-queue.json"

    static func fileURL() throws -> URL {
        let base = try FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let dir = base.appendingPathComponent("GroundworkPro", isDirectory: true)
        if !FileManager.default.fileExists(atPath: dir.path) {
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir.appendingPathComponent(fileName)
    }

    @objc func load(_ call: CAPPluginCall) {
        do {
            let url = try Self.fileURL()
            guard FileManager.default.fileExists(atPath: url.path) else {
                call.resolve(["value": NSNull()])
                return
            }
            let data = try Data(contentsOf: url)
            call.resolve(["value": String(data: data, encoding: .utf8) ?? ""])
        } catch {
            // A read failure must not break attendance: the JS layer falls back
            // to its localStorage mirror.
            call.resolve(["value": NSNull()])
        }
    }

    @objc func save(_ call: CAPPluginCall) {
        guard let value = call.getString("value") else {
            call.reject("value required")
            return
        }
        do {
            let url = try Self.fileURL()
            // Atomic so a kill mid-write cannot leave a truncated queue: the
            // write lands in a temp file and is renamed into place.
            try value.data(using: .utf8)?.write(to: url, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
            call.resolve()
        } catch {
            call.reject("Failed to persist attendance queue: \(error.localizedDescription)")
        }
    }

    @objc func clear(_ call: CAPPluginCall) {
        do {
            let url = try Self.fileURL()
            if FileManager.default.fileExists(atPath: url.path) {
                try FileManager.default.removeItem(at: url)
            }
            call.resolve()
        } catch {
            call.reject("Failed to clear attendance queue: \(error.localizedDescription)")
        }
    }
}
